### PR TITLE
PS-137: Doorman callfow

### PR DESCRIPTION
**Pull request by pavel-farkas**
New Doorman callflow developed for WhiteLabel Comms:
https://2600hz-commercial.atlassian.net/browse/PS-137

### CODE REVIEW

File: src/cf_exe.erl [hunk 1]
```diff
@@ -507,6 +507,18 @@ handle_cast({'channel_destroyed', JObj}, #state{stop_on_destroy='true'
...
+                                               ,call=Call
+                                               }=State) ->
+    case kapps_call:kvs_fetch('cf_last_action', Call) of
+        'cf_doorman' ->
+            lager:info("recv channel destroyed while cf_doorman is active CF module, going down"),
+            {'stop', 'normal', State#state{hangup_info=JObj}};
+        _ ->
+            lager:info("recv channel destroyed, noting but staying up"),
+            {'noreply', State#state{destroyed='true', hangup_info=JObj}}
+    end;
```
Comment (lazedo) [inline src/cf_exe.erl:521]: why are you trying to add it ? please remove handling specific modules on cf_exe

File: src/module/cf_doorman.erl [hunk 2]
```diff
@@ -290,11 +349,34 @@ create_callee_call(Data, CallerCall, Endpoints) ->
...
+    case kapps_call_command:b_channel_status(CalleeCall) of
+        {'ok', JObj} ->
+            case kz_json:get_boolean_value(<<"Answered">>, JObj) of
+                'true' ->
+                    lager:debug("Callee (B-leg) ~s has already answered. Terminating it.", [kapps_call:call_id_direct(CalleeCall)]),
+                    kapps_call_command:hangup(CalleeCall);
+                'false' ->
+                    UpdatedCalleeCall = kapps_call:from_json(JObj, CalleeCall),
+                    lager:debug("Callee (B-leg) ~s has not answered yet. Terminating it.", [kapps_call:call_id_direct(UpdatedCalleeCall)]),
+                    _ = freeswitch:api(binary_to_atom(kapps_call:switch_nodename(UpdatedCalleeCall)), 'uuid_kill', kapps_call:call_id_direct(UpdatedCalleeCall))
```
Comment (pavel-farkas) [inline src/module/cf_doorman.erl:375]: @jamesaimonetti I believe this is how we should not do it, but publish amqp request instead - however I didn't find how. It is kind of "special" situation: leg A is answered and parked, leg B is still ringing, and leg A terminates. At this moment Leg B is orphaned, it never learned about Leg A. 
I didn't find any other way how to terminate it - only by issuing direct 'uuid_kill' on the FS server which is dealing with the call.

File: src/cf_exe.erl [hunk 3]
```diff
@@ -507,6 +507,18 @@ handle_cast({'channel_destroyed', JObj}, #state{stop_on_destroy='true'
                                                }=State) ->
     lager:info("recv channel destroyed, going down"),
     {'stop', 'normal', State#state{hangup_info=JObj}};
+handle_cast({'channel_destroyed', JObj}, #state{stop_on_destroy='true'
+                                               ,cf_module_pid={_Pid, _Ref}
+                                               ,call=Call
+                                               }=State) ->
+    case kapps_call:kvs_fetch('cf_last_action', Call) of
+        'cf_doorman' ->
```
Comment (pavel-farkas) [inline src/cf_exe.erl:515]: New clause behaves differently only for cf_doorman module, as discussed. It won't affect any other CF module.

**Review by lazedo — CHANGES_REQUESTED** [review]
remove changes from `cf_exe`, as for the `cf_doorman` module, why are you making all these changes, might as well just create a new module, no ?

File: src/cf_exe.erl (hunk 3 above)
Comment (lazedo) [reply src/cf_exe.erl:515]: explain what you're trying to do and remove

File: src/module/cf_doorman.erl [hunk 4]
```diff
@@ -43,40 +45,69 @@
...
+    lager:debug("recording limit: ~p", [Limit]),
+    kz_math:clamp(16, 2, Limit).
 
-thank_caller(Data, Call) ->
-    Text = <<"Thanks!">>,
-    {'ok', _} = kapps_call_command:b_tts(Text, Call),
+prompt_caller(Call) ->
+    Text = <<"Hi there."
+        >>,
+    try
```
Comment (jamesaimonetti) [inline src/module/cf_doorman.erl:87]: Why the try/catch and not a simple case expression to handle the possible returns?

File: src/module/cf_doorman.erl [hunk 5]
```diff
@@ -290,11 +349,34 @@ create_callee_call(Data, CallerCall, Endpoints) ->
...
                ,{fun kapps_call:set_caller_id_name/2, kz_json:get_ne_binary_value(<<"caller_id_name">>, Data, <<"Doorman">>)}
                ,{fun kapps_call:set_caller_id_number/2, kapps_call:caller_id_number(CallerCall)}
                ,{fun kapps_call:set_custom_channel_var/3, <<"Retain-CID">>, <<"true">>}
-               ,{fun kapps_call:set_owner_id/2, kz_json:find(<<"owner_id">>, Endpoints)}
+            %    ,{fun kapps_call:set_owner_id/2, kz_json:find(<<"owner_id">>, Endpoints)}
                ,{fun kapps_call:set_resource_type/2, <<"audio">>}
                ],
     kapps_call:exec(Routines, kapps_call:new()).
+
+-spec handle_termination(kapps_call:call(), kz_json:object(), kz_json:object()) -> 'ok'.
```
Comment (jamesaimonetti) [inline src/module/cf_doorman.erl:362]: Third arg should be `kapps_call:call()` right?

File: src/module/cf_doorman.erl (hunk 2 above)
Comment (jamesaimonetti) [inline src/module/cf_doorman.erl:375]: You cannot assume that FreeSWITCH is connected to the same VM as the callflow process. Typically ecallmgr runs in its own VM and manages connections to FreeSWITCH, while callflows and other KAZOO apps run elsewhere with no connections to FreeSWITCH.

`kapps_call_command:hangup/1` should work here

**Review by jamesaimonetti — CHANGES_REQUESTED** [review]
First glance comments and suggestions

File: src/cf_exe.erl (hunk 1 above)
Comment (pavel-farkas) [reply src/cf_exe.erl:521]: I know this solution is not ideal. I need it to inform cf_doorman about termination of A leg before it gets bridged with B leg. There is a time period when both call legs are disconnected, while A leg is answered and parked and B leg is either ringing, or is answered and needs to pick one of the doorman menu options to proceed with call. When A leg is terminated during this time, I need some notification about it to properly handle B leg.

File: src/cf_exe.erl (hunk 3 above)
Comment (pavel-farkas) [reply src/cf_exe.erl:515]: Discussed with @jamesaimonetti about possible alternative solutions, I'll rewrite it and remove my code from cf_exe

**Pull request closed without merging**